text/3446-store.md

```diff
@@ -0,0 +1,1225 @@
+            return Err(AllocError);
+        }
+
+        let pointer = self.0.get() as *mut u8;
+
+        //  Safety:
+        //  -   Both starting and resulting pointers are in bounds of the same allocated objects as `old_layout` fits
+        //      `pointer`, as per the pre-conditions of `grow_zeroed`.
+        //  -   The offset does not overflow `isize` as `old_layout.size()` does not.
+        let pointer = unsafe { pointer.add(old_layout.size()) };
+
+        //  Safety:
+        //  -   `pointer` is valid, since `self` is valid.
+        //  -   `pointer` points to at an area of at least `mem::size_of::<T>() - old_layout.size()`.
+        //  -   Access to the next `mem::size_of::<T>() - old_layout.size()` bytes is exclusive.
+        unsafe { ptr::write_bytes(pointer, 0, mem::size_of::<T>() - old_layout.size()) };
+
+        Ok(((), mem::size_of::<T>()))
+    }
+}
+
+//  Safety:
+//  -   `self.resolve(handle)` always returns the same address, as long as `self` doesn't move.
+unsafe impl<T> StoreStable for InlineSingleStore<T> {}
+
+impl<T> fmt::Debug for InlineSingleStore<T> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> Result<(), fmt::Error> {
+        let layout = Layout::new::<T>();
+
+        f.debug_struct("InlineSingleStore")
+            .field("size", &layout.size())
+            .field("align", &layout.align())
+            .finish()
+    }
+}
+
+impl<T> InlineSingleStore<T> {
+    fn validate_layout(layout: Layout) -> Result<(), AllocError> {
+        let own = Layout::new::<T>();
+
+        if layout.align() <= own.align() && layout.size() <= own.size() {
+            Ok(())
+        } else {
+            Err(AllocError)
+        }
+    }
+}
+```
+
+And that's it!
+
+I need not implement `StoreMultiple`, and thus do not have to track allocations and deallocations. And I need not
+implement `StorePinning`, and thus do not have to ensure that memory remains pinned.
+
+More examples of `Store` can be found at https://github.com/matthieu-m/storage/tree/main/src/store, including an inline
+bump store.
+
+
+#   Reference-level explanation
+
+##  Overview
+
+This RFC introduces 5 new traits.
+
+The core of this RFC is the `Store` trait, a super-trait of the `StoreDangling` trait:
+
+```rust
+/// A trait allowing to get a dangling handle.
+pub unsafe trait StoreDangling {
+    /// Handle to a block of memory.
+    type Handle: Copy;
+
+    /// Returns a dangling handle.
+    ///
+    /// A dangling handle can be resolved to a pointer of at least the specified alignment, but the resulting pointer is
+    /// invalid
+    fn dangling(&self, alignment: Alignment) -> Result<Self::Handle, AllocError>;
+}
+
+/// Allocates and deallocates handles to blocks of memory, which can be temporarily resolved to pointers to actually
+/// access said memory.
+pub unsafe trait Store: StoreDangling {
+    /// Return a pointer to the block of memory associated to `handle`.
+    unsafe fn resolve(&self, handle: Self::Handle) -> NonNull<u8>;
```

Comment:
There seems to be a natural way to express the restrictions of `Store`, `StoreStable`, and `StorePinning` with an additional pointer type:

```rust
pub struct PointerGuard<T> {
    inner: NonNull<u8>,
    _guard: T,
}

impl<T> Deref for PointerGuard<T> {
    type Target = NonNull<u8>;
    fn deref(&self) -> &Self::Target { &self.inner }
}
```

- `Store::resolve` returns a pointer that is invalidated on allocation or move, so it could have the signature `unsafe fn resolve(&mut self, handle: Self::Handle) -> PointerGuard<&mut Self>`.
- `StoreStable` could provide a `resolve_stable` method with the signature `unsafe fn resolve_stable(&self, handle: Self::Handle) -> PointerGuard<&Self>`.
- `StorePinning` could provide a `resolve_pinning` method with the signature `unsafe fn resolve_pinning(&self, handle: Self::Handle) -> PointerGuard<Self>`.

The benefit of returning guards is catching easily-preventable misuses of the `Store` API. In concurrent situations, the default `Store::resolve` is unusable since threads may be calling allocation methods at random. And if you prevent them from doing so, then you've effectively done the work of materializing an `&mut MyStore` anyway. The same goes for `StoreStable` to prevent accidentally moving the store (for example, into a closure). `StorePinning` does neither but helps prevent the pointer from being used after the lifetime of the store expires. There are ways to circumvent this, but it would be a little helpful.

You mention later on that third-party libraries could fill this gap, but this seems like a pretty minimal add for some reasonable benefits. Any thoughts?

Reply:
It's worth noting that v2 made this distinction, but this revision deliberately collapses this all to a single `resolve` to reduce complex API surface, since these functions already necessarily have further `unsafe` usage restrictions, so while these specific ones can be modeled via lifetime, they're also just restrictions fairly in line with normal "unchecked reference" (pointer) requirements.

I think I'm in agreement now that for the specific case of this low-level interface, keeping all of the restrictions uniform as safety requirements serves to make the API simpler, and thus easier to use. While `PointerGuard` *would* make it slightly more difficult to accidentally misuse the API on that one axis, it does so at the cost of making the API as a whole more difficult to fit into your head.

Also, your `resolve_pinning` doesn't help, since it needs to clone the storage. Storage can be pinning without being sharable.

Reply:
Whoops, I meant to make `_guard` a `PhantomData<T>`. I've corrected my original comment; it should provide the desired guarantee without `clone` now.

I'm generally in agreement that the API surface should be minimal, and so if this discussion has already occurred then I'm happy.